# **Phase 2: Data Understanding**

In [1]:
# Loading file in Python for processing.
import pandas as pd
df = pd.read_csv('Churn_Modelling.csv')
df

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1


In [3]:
# Understanding the structure
print("Dataset Shape:", df.shape)

Dataset Shape: (10000, 14)


In [4]:
# Understanding data types and null values
print("--- Data Audit (Info) ---")
df.info()

--- Data Audit (Info) ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [5]:
# Statistical summary of numeric columns
df.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


## **Project Strategy: Core Metrics & Business Questions**
**Goal:** Analyze the banking dataset to identify the primary drivers of customer attrition (Churn) and provide data-driven retention strategies.

## 1. Core Project Metrics
We will focus on these foundational metrics to ensure a clear and accurate business story:
| **Metric** | **Purpose / Column Used** |
| :-----: | :-----: |
| **Total Churn Rate** | Measures the overall percentage of customers lost. (Exited) |
| **Geographic Risk** | Identifies which regions are `churn hotspots`. (Geography, Exited) |
| **Financial Vulnerability** | Evaluates if account balance size influences loyalty. (Balance, Exited) |
| **Engagement Impact** | Shows the correlation between bank activity and retention. (IsActiveMember, Exited) |
| **Credit Worthiness** | Analyzes if credit scores impact the decision to leave. (CreditScore, Exited) |
## 2. Key Analysis Questions
We will answer these specific questions to derive actionable business recommendations:
1. **Regional Impact:** Is the churn rate significantly higher in one specific country (e.g., Germany vs. France)?
2. **Financial Threshold:** Are low-balance customers more likely to churn, or is the bank losing high-value VIP clients?
3. **Product Stickiness:** Does owning multiple products (Credit Cards, etc.) increase a customer's "stickiness" to the bank?
4. **Member Activity:** Can increasing the engagement of "Inactive Members" effectively reduce churn?
5. **Demographic Trends:** Which age groups or genders show the highest risk of leaving the bank?

In [6]:
# Checking total number of rows.
print("Total Rows:", len(df))

Total Rows: 10000


In [7]:
# Checking Unique Customers.
print("Unique Customers:", df['CustomerId'].nunique())

Unique Customers: 10000


In [8]:
# Checking unique 'Surname' of customers.
print("Unique Surnames:", df['Surname'].nunique())

Unique Surnames: 2932


In [9]:
# Checking unique geographical areas.
print("Geographies Covered:", df['Geography'].unique())

Geographies Covered: ['France' 'Spain' 'Germany']


In [10]:
# Checking gender distribution.
print("Gender Distribution:\n", df['Gender'].value_counts())

Gender Distribution:
 Gender
Male      5457
Female    4543
Name: count, dtype: int64


## **Data Integrity & Structural Discovery**
## 1. Quantitative Audit (The "Grains" of Data)
Before starting the analysis, we performed a row-level audit to ensure the mathematical consistency of our dataset.
* **Dataset Volume:** The dataset contains `10,000` unique records.
* **Customer Uniqueness:** The count of Unique CustomerId (10,000) exactly matches the total number of rows. This confirms that there is no data duplication, and each row represents a single, unique banking customer.
* **Name Distribution:** There are `2,932` unique Surnames among the 10,000 customers. This suggests a high degree of diversity and indicates that the dataset is not biased towards a few specific families.
## 2. Categorical Distribution & Business Logic
We analyzed the primary dimensions of the bank's operational regions and its customer base:
* **Geographical Footprint:** The bank's data is concentrated in three European markets: `France`, `Spain`, and `Germany`. Our upcoming analysis will focus on which of these regions is the most "at-risk."
* **Gender Demographics:**
* **Male:** `5,457` customers
* **Female:** `4,543` customers
* **Analyst Insight:** The gender distribution is relatively balanced, with a slightly higher representation of male customers. This allows us to perform a fair gender-based churn comparison in the next phase.

In [11]:
# Saving data in file for future use.
df.to_csv('Churn_Modelling_for_Cleaning.csv')
print('File has been saved successfully.')

File has been saved successfully.
